### ISSUE TITLE
File stat fails when using a file name with characters with unicodes that don't fit on a single byte

### ISSUE THREAD

**Issue by Mogrinnar** [richgel999/miniz#75]
The character `Š` does not fit on a single byte. It's UTF-8 encoding is `197 160`, Unicode is `352`. Trying to fit this character in a `char` will result in 2 chars, `Å(197)` and `<No break space>1(60)` respectively.

The function `mz_zip_writer_add_file` takes a file name and a path, both are `const char *`. If the file name is `Š.txt` the function `mz_zip_get_file_modified_time` will return false because the `file_stat` will `==0`.

Since the file name is changed, when it is looked for it is not found and the file stat check fails.

Is miniz meant to work with file names with characters that have a unicode that are above 255 ?

**Comment by paulharris** [comment]
Just throwing it out there without looking, maybe it will help otherwise
please ignore me.
There have been unicode questions on other libraries recently, so I can't
remember where miniz is in terms of unicode filename support.

Using char* for utf8 is fine as long as the receiving API is going to
accept utf8.

Should work with linux since all filenames are just binary blobs to the OS,
and typically utf-8.

So I assume this problem is on Windows?
Need to use the W() (Unicode) API with UTF-16 text, ie convert before
passing it to any windows functions.
You usually can't use the standard posix API for unicode filenames, so
either use the widechar posix API that is on windows (and is often limited
in its abilities),
or use the standard W unicode functions like CreateFileW.

One final note, I've found it easier in the long run to use the \\?\ long
path names *anytime* I call a windows function.
It is fine (and often better) to use a plain utf8 encoded path right up to
the point where the windows function is called,
and then convert to a \\?\ long path in utf-16, then call the windows
function, and then discard the path.  ie only utf-16 at the point of the
call.
The path must be absolute, so relative paths must be converted to absolute
first (ie add the program's working directory at the front).

And also be careful not to convert a UNC path,
ie \\host\whatever should not become \\?\\\host\whatever

If miniz already does all that, then great :)  just ignore me.

**Comment by uroni** [comment]
You are correct. This needs special casing for Windows. Thanks for the bug report!

**Comment by jschmidt42** [comment]
I was able to work around this limitation by opening file handle myself and calling `mz_zip_writer_add_cfile` directly.

```cpp
auto dir = QDir(path);
QDirIterator it(dir, QDirIterator::Subdirectories);
while (it.hasNext()) {
auto entry = it.next();
auto relative_path = dir.relativeFilePath(entry);
auto file_info = QFileInfo(entry);
if (file_info.isFile()) {
	FILE* src_file = nullptr;
	fopen_s(&src_file, entry.toLocal8Bit(), "rb,ccs=UTF-8");
	if (!src_file) {
		mz_zip_writer_end(&zip);
		throw std::exception("Failed to open file to be zipped");
	}

	time_t ft = file_info.created().toTime_t();
	if(!mz_zip_writer_add_cfile(&zip, relative_path.toUtf8(), src_file, file_info.size(), &ft, nullptr, 0, level_and_flags, nullptr, 0, nullptr, 0)) {
		fclose(src_file);
		mz_zip_writer_end(&zip);
		throw std::exception(mz_zip_get_error_string(mz_zip_get_last_error(&zip)));
	}
	fclose(src_file);
}
```

**Comment by paulharris** [comment]
I don't understand why you would want to use the ccs=utf-8 flag in fopen,
especially with the binary 'b' flag.
That will autoconvert the text inside the file from utf-8 to UTF-16... not
what you want for a zip file.  It probably only works because the binary
flag disables the encoding mode.
The ccs flag has nothing to do with filenames.

fopen() will not work with unicode filenames on Windows.

Convert the filenames to UTF-16 and use _wfopen()

Also, plain "C:\filename.txt" paths will not work if the path is long.
You will need to convert it to \\?\c:\filename.txt
ie a UNC path, if you want it to work for long paths (ie full path longer
than 255 chars).
Note that UNC paths must be absolute, no relative paths.
Note that a \\host\share\file.txt network path should not be converted to
UNC.
There are more details but that will work.

So use absolute UNC UTF-16 paths on windows with _wfopen().

**Comment by jschmidt42** [comment]
@paulharris thanks for the feedback. Actually, instead of using miniz high level functions, we now do our own file open and use lower level of miniz and providing our read and write callbacks and now everything seems to work fine and we are able to handle all UTF-8 cases.

Thanks,

**Comment by paulharris** [comment]
Ok no worries

**Comment by uroni** [comment]
Windows 10 build 17134 got a beta option "Beta: Use Unicode UTF-8 for worldwide language support". So this might actually fix itself with newer Windows versions.

**Comment by chchwy** [comment]
I have a solution for Unicode filenames on Windows.
Unfortunately, this solution needs C++ headers.

1.  Convert the filename string to UTF-8 when you call `mz_zip_reader_init_file()`
For example with Qt, you can do 
```cpp
// zipFilePath is a QString
mz_bool ok = mz_zip_reader_init_file(mz, zipFilePath.toUtf8().data(), 0);
```

2. Go to miniz source code, replace the default `mz_fopen()` with the following one:
```cpp
#include <codecvt>
#include <string>
#include <locale>

static FILE *mz_fopen(const char *pFilename, const char *pMode)
{
    // convert UTF-8 to wide characters 
    std::wstring_convert<std::codecvt_utf8_utf16<wchar_t>> converter;
    std::wstring sWideFilename = converter.from_bytes(pFilename);
    std::wstring sWideMode = converter.from_bytes(pMode);

    FILE *pFile = NULL;
    _wfopen_s(&pFile, sWideFilename.c_str(), sWideMode.c_str());
    return pFile;
}
```
The key point is using `_wfopen_s()` instead of `fopen()`. Then miniz should be able to handle Unicode filenames. You may want to do the same trick for `mz_freopen()` as well. Just convert the UTF-8 path to `wchar_t` and then use the wide-character version API to open a file.

I have been using this trick in my project [Pencil2D](https://github.com/pencil2d/pencil) and it's working fine since 2018.

**Comment by paulharris** [comment]
https://docs.microsoft.com/en-us/windows/uwp/design/globalizing/use-utf8-code-page

If you limit your app to only new Windows 10 (after May 2019 update), then you could use Window's new UTF-8 code page - add to your program's manifest.

I wish this had come a LONG time ago.
